File: cpp/Ice/middleware/AuthorizationMiddleware.cpp
// Copyright (c) ZeroC, Inc.

#include "AuthorizationMiddleware.h"

#include <iostream>

using namespace std;

Server::AuthorizationMiddleware::AuthorizationMiddleware(Ice::ObjectPtr next, string validToken)
    : _next{std::move(next)},
      _validToken{std::move(validToken)}
{
}

void
Server::AuthorizationMiddleware::dispatch(
    Ice::IncomingRequest& request,
    function<void(Ice::OutgoingResponse)> sendResponse)
{
    // Check if the request has a valid token in its context.
    auto p = request.current().ctx.find("token");
    string token = p != request.current().ctx.end() ? p->second : "";
    if (token != _validToken)
    {
        cout << "Rejecting request with invalid token '" << token << "'" << endl;
        throw Ice::DispatchException{
            __FILE__,
            __LINE__,
            Ice::ReplyStatus::Unauthorized,
            "Invalid token '" + token + "'"};
    }

    cout << "Accepting request with token '" << token << "'" << endl;

    // Continue the dispatch pipeline.
    _next->dispatch(request, std::move(sendResponse));
}

File: swift/Ice/middleware/Sources/Server/AuthorizationMiddleware.swift
// Copyright (c) ZeroC, Inc.

import Ice

/// The AuthorizationMiddleware intercepts incoming requests and checks the token entry in their contexts.
final class AuthorizationMiddleware: Ice.Dispatcher {
    private let next: Dispatcher
    private let validToken: String

    /// Initializes an AuthorizationMiddleware.
    /// - Parameters:
    ///   - next: The next dispatcher in the dispatch pipeline.
    ///   - validToken: The value to check against.
    init(next: Dispatcher, validToken: String) {
        self.next = next
        self.validToken = validToken
    }

    /// Dispatches the incoming request, checking for a valid token.
    func dispatch(_ request: IncomingRequest) async throws -> OutgoingResponse {
        let token = request.current.ctx["token"] ?? ""

        guard token == validToken else {
            print("Rejecting request with invalid token '\(token)'")
            throw DispatchException(
                replyStatus: ReplyStatus.unauthorized.rawValue, message: "Invalid token '\(token)'")
        }

        print("Accepting request with token '\(token)'")

        // Continue the dispatch pipeline.
        return try await next.dispatch(request)
    }
}
